x/mongo/driver/topology/server.go

```diff
@@ -755,32 +788,44 @@ func copyConnectionOpts(opts []ConnectionOption) []ConnectionOption {
 	return optsCopy
 }
 
-func (s *Server) setupHeartbeatConnection() error {
+func (s *Server) setupHeartbeatConnection(ctx context.Context) error {
 	conn := s.createConnection()
 
-	// Take the lock when assigning the context and connection because they're accessed by cancelCheck.
+	// Take the lock when assigning the connection because they're accessed by
+	// cancelCheck.
 	s.heartbeatLock.Lock()
-	if s.heartbeatCtxCancel != nil {
-		// Ensure the previous context is cancelled to avoid a leak.
-		s.heartbeatCtxCancel()
-	}
-	s.heartbeatCtx, s.heartbeatCtxCancel = context.WithCancel(s.globalCtx)
+
 	s.conn = conn
+
+	if s.cfg.connectTimeout != 0 {
+		var cancelFn context.CancelFunc
+		ctx, cancelFn = context.WithTimeout(ctx, s.cfg.connectTimeout)
+
+		defer cancelFn()
+	}
+
 	s.heartbeatLock.Unlock()
 
-	return s.conn.connect(s.heartbeatCtx)
+	return s.conn.connect(ctx)
 }
 
-// cancelCheck cancels in-progress connection dials and reads. It does not set any fields on the server.
+// cancelCheck cancels in-progress connection dials and reads. It does not set
+// any fields on the server.
 func (s *Server) cancelCheck() {
 	var conn *connection
 
-	// Take heartbeatLock for mutual exclusion with the checks in the update function.
+	// Take heartbeatLock for mutual exclusion with the checks in the update
+	// function.
 	s.heartbeatLock.Lock()
-	if s.heartbeatCtx != nil {
-		s.heartbeatCtxCancel()
-	}
+
+	s.cancelHeartbeatCheckOnce.Do(func() {
+		s.cancelHeartbeatCheck <- struct{}{}
+
+		s.heartbeatCheckCanceled = true
+	})
```

Comment:
What happens the 2nd time that `cancelCheck` is called if this block is only run once? It seems like it won't cancel a heartbeat check anytime after the first one is canceled.

Reply:
Great catch. I looked at various patterns to avoid keeping the context on a struct, including [this](https://cs.opensource.google/go/go/+/refs/tags/go1.22.3:src/context/context.go;l=413) `closedchan` pattern used by the `cancelCtx` (which doesn't work because it assumes cancel can only be called once). I propose using a singleton to cancel the current check, reset with each iteration:

![heartbeat_cancel_flowchart](https://github.com/mongodb/mongo-go-driver/assets/24281431/446e3a2c-ac16-48f3-9b4d-004406a61069)

Reply:
I believe this has been resolved by the check-cancel logic refactor using the `contextListener`.